contracts/DNSSECImpl.sol

```diff
@@ -238,30 +209,139 @@ contract DNSSECImpl is DNSSEC, Owned {
+        RRUtils.NSEC3 memory ceNSEC3 = readNSEC3(ce);
+        // The flags field must be 0 or 1 (RFC5155 section 8.2).
+        require(ceNSEC3.flags & 0x01 == 0);
+        // Check that the closest encloser is from the correct zone (RFC5155 section 8.3)
+        // "The DNAME type bit must not be set and the NS type bit may only be set if the SOA type bit is set."
+        require(!ceNSEC3.checkTypeBitmap(DNSTYPE_DNAME) && (!ceNSEC3.checkTypeBitmap(DNSTYPE_NS) || ceNSEC3.checkTypeBitmap(DNSTYPE_SOA)));
+
+        // Case 1: deleteName does exist, but no records of RRTYPE deleteType do.
+        if(isMatchingNSEC3Record(deleteType, deleteName, ce.name, ceNSEC3)) {
+            delete rrsets[keccak256(deleteName)][deleteType];
+        // Case 2: deleteName does not exist.
+        } else if(isCoveringNSEC3Record(deleteType, deleteName, ce.name, ceNSEC3, nc.name, readNSEC3(nc))) {
+            delete rrsets[keccak256(deleteName)][deleteType];
         } else {
-            // deleteName must come after nsecName
-            require(deleteNameHash > nsecNameHash);
-            // Check if the NSEC next name comes after the NSEC name.
-            if(nextNameHash > nsecNameHash) {
-                // deleteName must come also come before nextName
-                require(deleteNameHash < nextNameHash);
+            revert();
+        }
+    }
+
+    function checkNSEC3Validity(RRUtils.SignedSet memory nsec, bytes memory deleteName, int32 originalInception) private pure {
+        // The records must have been signed after the record we're trying to delete
+        require(int32(nsec.inception) - originalInception >= 0);
+
+        // The record must be an NSEC3
+        require(nsec.typeCovered == DNSTYPE_NSEC3);
+
+        // nsecName is of the form <hash>.zone.xyz. <hash> is the NSEC3 hash of the entire name the NSEC3 record matches, while
+        // zone.xyz can be any ancestor of that name. We'll check that, so someone can't use a record on foo.com
+        // as proof of the nonexistence of bar.org.
+        require(checkNSEC3OwnerName(nsec.name, deleteName));
+    }
+
+    function isMatchingNSEC3Record(uint16 deleteType, bytes memory deleteName, bytes memory closestEncloserName, RRUtils.NSEC3 memory closestEncloser) private view returns(bool) {
+        // Check the record matches the hashed name, but the type bitmap does not include the type
+        if(checkNSEC3Name(closestEncloser, closestEncloserName, deleteName)) {
+            return !closestEncloser.checkTypeBitmap(deleteType);
+        }
+
+        return false;
+    }
+
+    function isCoveringNSEC3Record(uint16 deleteType, bytes memory deleteName, bytes memory ceName, RRUtils.NSEC3 memory ce, bytes memory ncName, RRUtils.NSEC3 memory nc) private view returns(bool) {
+        // The flags field must be 0 or 1 (RFC5155 section 8.2).
+        require(nc.flags & 0x01 == 0);
```

Comment:
Does the code fit the comment?  Seems like the bitmask should be 0xfe here.

Reply:
Good point. Fixed!

Reply:
Good point. Fixed!